### ISSUE TITLE
noise ceiling calculation

### ISSUE THREAD

**Issue by sklin93** [kendrickkay/nsddatapaper#1]
Hi,

I run the [noise ceiling script](https://github.com/cvnlab/nsddatapaper/blob/a951ce8568e2691d69949925e836a4b389f42aa1/main/export_noiseceiling.m#L75)  on 1.8mm b3 betas for subject 1, and got the [ncsnr](https://imgur.com/a/LDTmKwt) (median 0.0680, mean 0.0873) way smaller than the 0.260 that the paper indicates. Even though it's only on one subject instead of multiples, it shouldn't differ that much?

I also tried to extract visual cortex voxels with prf-visualrois, and the ncsnr will be increased to median 0.1165, mean 0.1558 (still much lower than 0.260).
I used the same calculation on bold5k's visual cortex voxels, and the median and mean ncsnr are both around 0.6735.

I'm really confused about the results and don't understand what went wrong...any suggestions?

**Comment by kendrickkay** [comment]
Yes, what you are describing doesn't sound quite right.  A few ideas:
-Are you more or less running the script as-is, or did you tweak/modify anything?
-Can you compare your results to the provided ncsnr.nii.gz files, to help diagnose what's going on?
-Certainly the ncsnr values vary dramatically across the brain, so you'll want to be sensitive to that issue (which it sounds you are)
-Are you making sure to z-score each voxel's betas within each session, and then aggregating across sessions?
-Some screenshots or some example code that you are looking at would be helpful for diagnosis purposes.

**Comment by sklin93** [comment]
Thank you for the reply!

- I'm running the script as-is, only changing the directory, session number (since only 37 sessions are available), and subject number (I'm only running for subject 1). You can see all the code changes [here](https://github.com/sklin93/nsddatapaper/commit/e4f1b506b2cdd78b62b3efec0a7005e179009575#diff-62ff4956a4e08ff7d85a1b1b811a861dd352c7c47d251f127ded7bcd50e23feb). However, I did also write a python version based on the algorithm (got the same result as this Matlab script for all voxel ncsnr) for visual cortex ncsnr and bold5k ncsnr.
- I downloaded the corresponding file: https://natural-scenes-dataset.s3.amazonaws.com/nsddata_betas/ppdata/subj01/func1pt8mm/betas_fithrf_GLMdenoise_RR/ncsnr.nii.gz and checked with [these codes](https://imgur.com/a/9Z5SO8g), and got `median 0.06803017109632492 mean 0.08819003264845511 max 1.418752908706665`, which are quite close to what I got from the Matlab script, but shouldn't it be much higher?
- Was the 0.26 median ncsnr obtained with any specific ROI mask?
- I think the Matlab script is zscore-ing each voxel within each session (`betas = calczscore(betas,4,[],[],0);`) so it shouldn't be the reason? I also made sure to do the same when I'm using my python script.
- Here's my python script, but my main issue is getting the same small ncsnr with the original Matlab version:
```
''' n is repetition number, n=3 for nsd
    sig is a dictionary with keys being image_ids and values being betas (3, voxel_num),
    these betas are already voxel-wise z-scored per session.
'''

_fmri_var = []
for k, v in tqdm(sig.items()):
    if len(v) == n:  # ignore images only repeated 2 times in the available data
        _fmri_var.append(np.var(v, 0, ddof=1))

noise_std = np.sqrt(np.stack(_fmri_var).mean(0)) # for every voxel, shape is (voxel_num,)
print(f'median noise_std: {np.median(noise_std)}',
      f'mean noise_std: {noise_std.mean()}, max noise_std: {max(noise_std)}')

signal_var = 1 - noise_std**2
'''based on https://github.com/cvnlab/nsddatapaper/blob/main/main/translatevmetric.m,
   only keep the positive part'''
signal_std = np.sqrt(np.maximum(0, signal_var))
ncsnr = signal_std / noise_std
print(f'median ncsnr: {np.median(ncsnr)}, mean ncsnr: {ncsnr.mean()}',
      f'max ncsnr: {max(ncsnr)}')

'''Noice ceiling'''
nc = ncsnr**2 / (ncsnr**2 + 1.0/n)
print(f'median nc: {np.median(nc)}, mean nc: {nc.mean()}, max nc: {max(nc)}')
```

(and just in case you missed the file, there's no nsd_savenifti.m available and I also couldn't find it elsewhere like calczscore, chunking, etc.)
Thanks again!

**Comment by kendrickkay** [comment]
This is informative. A few thoughts:
- since you have access to only 37, the numbers may differ slightly. (Also, note that you won't be able to get the full triplets (3 trials) for all images.)
- ncsnrs vary dramatically across the brain. this is expected since this is a visual/memory experiment!
- yes, the summary noise ceiling levels (which are a simple transformation of ncsnr) are reported within the nsdgeneral ROI (see attached screenshot)
- and yes, a lot of the code is just in matlab, so you'll have to find/implement python equivalents.

So my guess is that everything is still correct; you just need to do voxel selection (e.g. according to the nsdgeneral ROI)

![image](https://user-images.githubusercontent.com/2728338/149956336-8caacbdc-a315-42d1-8e20-a9d6c8316284.png)

**Comment by sklin93** [comment]
Thank you!
- I checked with the provided ncsnr.nii.gz (under `func1pt8mm/betas_fithrf_GLMdenoise_RR` folders of all subjects) and provided nsdgeneral.nii.gz (under `func1pt8mm/roi` folders of all subjects) with the following code:
```
for subj in range(1, 9):
    nii_data = nib.load(data_dir + f'nsd/ncsnr-subj{subj}.nii.gz').get_fdata()
    mask = nib.load(data_dir + f'nsd/nsdgeneral-subj{subj}.nii.gz').get_fdata()
    mask = (mask != -1)
    nii_data = nii_data[mask]

    print(f'subject {subj} nsdgeneral voxel number: {nii_data.shape}')
    print(f'ncsnr median: {np.median(nii_data): .4f}, mean: {np.mean(nii_data): .4f},',
          f'max:{np.max(nii_data): .4f}')
```
And here's what I got:
```
subject 1 nsdgeneral voxel number: (107104,)
ncsnr median:  0.1219, mean:  0.1637, max: 1.4188
subject 2 nsdgeneral voxel number: (107860,)
ncsnr median:  0.1196, mean:  0.1654, max: 1.3350
subject 3 nsdgeneral voxel number: (111068,)
ncsnr median:  0.0916, mean:  0.1201, max: 1.2693
subject 4 nsdgeneral voxel number: (101628,)
ncsnr median:  0.0963, mean:  0.1226, max: 1.3528
subject 5 nsdgeneral voxel number: (94578,)
ncsnr median:  0.1174, mean:  0.1685, max: 1.6371
subject 6 nsdgeneral voxel number: (120668,)
ncsnr median:  0.1338, mean:  0.1606, max: 1.1553
subject 7 nsdgeneral voxel number: (90214,)
ncsnr median:  0.1014, mean:  0.1281, max: 1.1764
subject 8 nsdgeneral voxel number: (108767,)
ncsnr median:  0.0736, mean:  0.0897, max: 1.1062 
```
It seems I still can't get the "median ncsnr averaged across the first four NSD participants being 0.260" using the provided ncsnr. Any ideas..?

- And for the bold5k dataset, what preprocessing was done before its ncsnr calculation (0.187 as suggested in the nsd paper)? I z-scored the signals voxel-wise per session, only using visual cortex voxels (ROI provided in bold5k dataset), and averaged signal of TR3 and TR4 as suggested by its paper. I also only used images that have 4 repetitions. I got quite high ncsnr for their data;  am I missing something during the processing?

**Comment by kendrickkay** [comment]
Aha, I think I see what's happening.  The analysis sentence described in the paper "median ncsnr averaged..." was done using a slighlty different pipeline (spearheaded by Jacob Prince).  The main noise ceiling results provided with NSD data release (and shown in Fig 3) are somewhat different.  

The issues at stake here are tricky spatial issues like, which voxels are included, whether the data preparation is volume or surface, which voxels are deemed to be "gray matter" and things like that.  Jacob analyzed BOLD5000 in addition to NSD, and the preparation of BOLD5000 differed, potentially, from what you are looking at.  In particular, he re-ran the exact same GLM methods (as opposed to a TR-by-TR analysis) on the BOLD5000 data.  In directly comparing NSD and BOLD5000, a "MNI space" nsdgeneral ROI was made to ensure fair comparison (as described in the paper).  If you really want to reproduce this side analysis, we'd have to dig up the MNI ROI mask and probably compare directly to Jacob's analysis and have a Zoom discussion with him.  But I'm not sure if that's your real goal here?

If your goal is to double check your noise ceiling analysis, i think the export_noiseceiling.m script and re-implementing the noise ceiling analysis and comparing to the provided ncsnr.nii.gz files is your best bet, and it sounds like you've done that.

**Comment by sklin93** [comment]
Thank you, this makes sense now! 

My actual goal is to decide which dataset to use for reconstructing images from fMRI signals, and that means choosing a signal source with higher ncsnr (of course also the more samples the better).

My dilemma was BOLD5000 (TR averaged, non-beta) has a better median ncsnr than NSD beta3 while NSD has more data, and annotations are easier to get from the COCO API. So it would be great if I could confirm I calculated something wrong and NSD in fact has a better ncsnr than BOLD5000 -- then I can just go ahead and use NSD.

But it seems with this particular noise ceiling calculation and the data I have on hand (BOLD5000 TR, NSD beta), the NSD beta still has a worse ncsnr...Or maybe this isn't a fair comparison at all?

**Comment by kendrickkay** [comment]
I see. Based on our extensive look at BOLD5000 and NSD, we are fairly confident NSD is overall much better than BOLD5000, but it is good I suppose to double check.

Aside from tricky voxel selection issues (like discussed earlier), I suppose another thing to check is your handling of the BOLD5000 data.  Are you properly extracting/handling the few repetitions that are available with that dataset?  Also, possibly there might be an issue in how you are normalizing and combining the TR-data across sessions?

**Comment by sklin93** [comment]
Yes, BOLD5000 has 112 images repeated four times for subject-1,2 and 107 images repeated four times for subject-3. I checked the image number and names, making sure I only used these images for the ncsnr calculation of each subject. 

~~For the signal, I just voxelwise z-scored their pre-processed signals (fmriprep derivative), applied their provided visual cortex mask (spm derivative), then averaged TR3 and TR4's signals since its paper has a figure showing subject-1,2,3's responses all peak around these two TRs).~~
I realized my mistake in the calculation while I'm writing this! The standardized-then-averaged TR3.5 doesn't have variance 1 on each voxel...I plugged in their actual variances instead of 1 when subtracting the noise variance, and now the ncsnr seems much normal (median 0.1206, mean 0.14856, max 0.7279).

Thanks so much for pointing it out!!! I can finally fully convince myself to use NSD :D
